GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,216 @@
+        {
+            Unknown = 0,
+            Stopped = 1,
+            Failing = 2,
+            Healthy = 3,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.isRunning)
+            {
+                return;
+            }
+
+            this.isRunning = true;
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeStringWriter)} is not started");
+            }
+
+            return this.queue.TryAdd(message);
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            this.isRunning = false;
+            this.queue.CompleteAdding();
+            this.writerThread.Join();
+
+            Debug.Assert(this.queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public void Dispose()
+        {
+            if (!this.isDisposed)
+            {
+                this.Stop();
+
+                this.pipeClient?.Dispose();
+                this.pipeClient = null;
+
+                this.writerThread = null;
+            }
+
+            this.isDisposed = true;
+        }
+
+        private void RaiseStateChanged(QueuedPipeStringWriterState newState, Exception ex)
+        {
+            int prevState = Interlocked.Exchange(ref this.state, (int)newState);
+
+            if (prevState != (int)newState)
+            {
+                this.StateChanged?.Invoke(this, new QueuedPipeStringWriterStateEventArgs(newState,  ex));
+            }
+        }
+
+        private void ThreadProc()
+        {
+            // Drain the queue of all messages currently in the queue
+            do
+            {
+                try
+                {
+                    // Take() is a blocking call that will return a message when there is one available
+                    // or will throw an InvalidOperationException when the queue has been marked as completed.
+                    string message = this.queue.Take();
+                    if (message != null)
```

Comment:
Should null messages ever be put on the queue?  I guess since it's part of the logger itself it would hard to log that condition.

Reply:
That's just going to be a limitation I'm enforcing here. It makes no sense to log null. Think about what we'd send over the pipe? Zero bytes.